optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java

```diff
@@ -53,31 +62,28 @@
      * Maximum capacity: The maximum capacity for each resource for each machine must not be exceeded.
      */
     private Constraint maximumCapacity(ConstraintFactory factory) {
-        throw new UnsupportedOperationException("Not yet implemented due to missing support for bi-grouping.");
-//        return factory.from(MrMachineCapacity.class)
-//                .join(MrProcessAssignment.class,
-//                        equal(MrMachineCapacity::getMachine, MrProcessAssignment::getMachine)
-//                )
-//                .groupBy(
-//                        (machineCapacity, processAssignment) -> machineCapacity, sumLong(
-//                                (machineCapacity, processAssignment) -> processAssignment.getUsage(machineCapacity.getResource())
-//                        )
-//                )
-//                .filter(((machineCapacity, usage) -> machineCapacity.getMaximumCapacity() < usage))
-//                .penalizeLong(MrConstraints.MAXIMUM_CAPACITY,
-//                        HardSoftLongScore.ofHard(1L),
-//                        (machineCapacity, usage) -> machineCapacity.getMaximumCapacity() - usage);
+        return factory.from(MrMachineCapacity.class)
+                .join(MrProcessAssignment.class,
+                      equal(MrMachineCapacity::getMachine, MrProcessAssignment::getMachine)
+                )
+                .groupBy((machineCapacity, processAssignment) -> machineCapacity, sumLong(
+                        (machineCapacity, processAssignment) -> processAssignment.getUsage(machineCapacity.getResource()))
+                )
+                .filter(((machineCapacity, usage) -> machineCapacity.getMaximumCapacity() < usage))
+                .penalizeLong(MrConstraints.MAXIMUM_CAPACITY,
+                              HardSoftLongScore.ofHard(1L),
+                              (machineCapacity, usage) -> machineCapacity.getMaximumCapacity() - usage);
```

Comment:
If the usage is higher than the capacity, doesn't `capacity - usage` return a negative number? But we have a fail fast for that.
What am I missing?

Reply:
Good catch; there was no machine breaking this constraint. Once I removed the filter, the fast fast worked.